pkg/kube-metrics/collector.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubemetrics
+
+import (
+	"context"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	kcollector "k8s.io/kube-state-metrics/pkg/collector"
+	"k8s.io/kube-state-metrics/pkg/metric"
+	metricsstore "k8s.io/kube-state-metrics/pkg/metrics_store"
+)
+
+// NewCollectors returns collections of metrics in the namespaces provided, per the api/kind resource.
+// The metrics are registered in the custom generateStore function that needs to be defined.
+func NewCollectors(cfg *rest.Config, namespaces []string, api string, kind string, metricFamily []metric.FamilyGenerator) ([]kcollector.Collector, error) {
+	namespaces = deduplicateNamespaces(namespaces)
+	var collectors []kcollector.Collector
+	// Generate collector per namespace.
+	for _, ns := range namespaces {
+		dclient, err := newClientForGVK(cfg, api, kind)
```

Comment:
My thought was actually to call `newClientForGVK` in ServeCRMetrics, and pass the `dynamic.NamespaceableResourceInterface` into `NewCollectors` instead of the `rest.Config`. The big reason I'm suggesting that is because I was doing something similar with the `rest.Config` in the Helm role generation and realized I couldn't test it easily. Using an interface instead made it much easier.

I think that would also mean that `NewCollectors` would no longer need to return an error, which would be nice.

Reply:
done 👍